os/mm/mm_heap/mm_heapinfo_group.c

```diff
@@ -0,0 +1,189 @@
+#include <tinyara/config.h>
+#include <tinyara/sched.h>
+#include <tinyara/mm/mm.h>
+#include <stdint.h>
+#include <string.h>
+#include <tinyara/mm/heapinfo_internal.h>
+
+int max_group;
+struct heapinfo_group_s heapinfo_group[HEAPINFO_USER_GROUP_NUM];
+struct heapinfo_group_info_s group_info[HEAPINFO_THREAD_NUM];
+
+/****************************************************************************
+ * Name: heapinfo_update_group
+ *
+ * Description:
+ * Update Peak heap size for Group
+ ****************************************************************************/
+void heapinfo_update_group(mmsize_t size, pid_t pid)
+{
+	int check_idx;
+	int group_num;
+	int stack_pid;
+	for (check_idx = 0; check_idx < HEAPINFO_THREAD_NUM; check_idx++) {
+		if (pid == group_info[check_idx].pid) {
+			group_num = group_info[check_idx].group;
+			heapinfo_group[group_num].curr_size += size;
+			/* Update peak size */
+			if (heapinfo_group[group_num].curr_size > heapinfo_group[group_num].peak_size) {
+				heapinfo_group[group_num].peak_size = heapinfo_group[group_num].curr_size;
+				/* calculate the summation of stacks */
+				heapinfo_group[group_num].stack_size = 0;
+				for (stack_pid = 0; stack_pid < HEAPINFO_THREAD_NUM; stack_pid++) {
+					if (group_info[stack_pid].pid != -1 && group_info[stack_pid].group == group_num) {
+						heapinfo_group[group_num].stack_size += group_info[stack_pid].stack_size;
+					}
+				}
+				heapinfo_group[group_num].heap_size = heapinfo_group[group_num].peak_size - heapinfo_group[group_num].stack_size;
+			}
+			break;
+		}
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_update_group_info
+ *
+ * Description:
+ * when create or release task/thread, check that the task/thread is 
+ * in group list
+ ****************************************************************************/
+void heapinfo_update_group_info(pid_t pid, int group, int type)
+{
+	int info_idx;
+	struct tcb_s *tcb;
+	switch (type) {
+	case HEAPINFO_INIT_INFO:
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			group_info[info_idx].pid = pid;
+			group_info[info_idx].group = group;
+			group_info[info_idx].stack_size = 0;
+		}
+		break;
+	case HEAPINFO_ADD_INFO:
+		if (group > max_group) {
+			max_group = group;
+		}
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			if (group_info[info_idx].pid <= 0) {
+				group_info[info_idx].pid = pid;
+				group_info[info_idx].group = group;
+				tcb = sched_gettcb(pid);
+				if (tcb) {
+					group_info[info_idx].stack_size = tcb->adj_stack_size;
+					heapinfo_update_group(tcb->adj_stack_size, pid);
+				}
+				break;
+			}
+		}
+		break;
+	case HEAPINFO_DEL_INFO:
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			if (pid == group_info[info_idx].pid) {
+				heapinfo_update_group((-1) * group_info[info_idx].stack_size, pid);
+
+				group_info[info_idx].pid = -1;
```

Comment:
Yes. like this, this is common style of init / deinit.

Reply:
Thanks.